rfd/0078-role-variables.md

```diff
@@ -0,0 +1,358 @@
+mapping per group.
+
+Solution today:
+
+```yaml
+kind: role
+version: v5
+metadata:
+  name: users
+spec:
+  allow:
+    node_labels:
+      env:
+        - regexp.replace(external.groups, "dev", "dev")
+        - regexp.replace(external.groups, "dev", "staging")
+        - regexp.replace(external.groups, "qa", "qa")
+        - regexp.replace(external.groups, "qa", "staging")
+    app_labels:
+      env:
+        - regexp.replace(external.groups, "dev", "dev")
+        - regexp.replace(external.groups, "dev", "staging")
+        - regexp.replace(external.groups, "qa", "qa")
+        - regexp.replace(external.groups, "qa", "staging")
+```
+
+With more roles and more complex rules, this would be even more difficult. The
+full list would have to be repeated for every value in every role. I don't think
+anyone even knows you can do this, and I hope no-one actually does, because it's
+ridiculous.
+
+With role variables it becomes trivial:
+
+```yaml
+kind: role
+version: v5
+metadata:
+  name: users
+spec:
+  vars:
+    - name: allow_envs
+      values:
+        - input: "{{external.groups}}"
+          match: "^(qa|devs)$"
+          out:
+            - "$1"
+            - "staging"
+  allow:
+    node_labels:
+      env: "{{vars.allow_envs}}"
+    app_labels:
+      env: "{{vars.allow_envs}}"
+```
+
+The same variable can easily be defined once, and used many times within this
+role.
+
+## Alternatives
+
+### Trait transformations
+
+Instead of defining variables in the role spec, we could create a new `trait`
+resource which can define new traits which will be computed during login and
+embedded in the user's cert.
+
+These traits would be reusable across roles and trusted clusters.
+
+The syntax could be similar to the proposed role variable syntax:
+
+```yaml
+kind: trait
+version: v1
+metadata:
+  name: groups
+spec:
+  values:
+    # An input will expand to a list of N inputs if the trait is a list of N.
+    # Any input value which matches the `match` regex will contribute one
+    # output value, which may contain regex captures from that specific input.
+    #
+    # Here [devs, env-staging, env-prod] maps to [staging, prod].
+    - input: "{{external.groups}}"
+      match: '^env-(\w+)$'
+      out: ['$1']
+    # If any of the N input values match the `not_match` regex, none of the
+    # values will be used.
+    #
+    # Here [devs] maps to [dev], but [devs, contractors] maps to []
+    - input: "{{external.groups}}"
+      match: '^devs$'
+      not_match: 'contractors'
+      out: ["dev"]
+```
+
+### Common Expression Language
+
+Another option for defining role variables (or traits) would be to use an
+expression language rather than the proposed regular expression based
+construction.
+
+Rewritten to use
+[CEL](https://github.com/google/cel-go)
+the first example in this RFD could look like:
+
+```yaml
+kind: role
+version: v5
+metadata:
+  name: node_users
+spec:
+  vars:
+    - name: "logins"
```

Comment:
Let's not introduce a language that is almost like predicate, but a bit different. Instead, let's focus on investing in predicate.